### DISCUSSION TITLE
Embedded `caf` audio files do not play in all browsers

### DISCUSSION THREAD

**Discussion by ldexterldesign** [ReagentX/imessage-exporter#371] [Q&A] [answered]
👋

Hope you're OK!

Thanks for awesome software 😍

Export of voice/audio/.caf files is successful as attachments but the [audio] elems don't seem to play 😕 - I presume you're aware of [this] (although issue is closed and last/latest comment means nothing to me 🤔) but what's the reason and will it get fixed/enhanced soon?

Hope to hear back

Kind thanks

[audio]: https://devdocs.io/html/element/audio
[this]: https://github.com/ReagentX/imessage-exporter/issues/155

**Comment by ReagentX** [comment]
Can you post a screenshot of what your browser displays?

What browser are you using? Most browsers don't support that format: https://en.wikipedia.org/wiki/HTML_audio#Supported_audio_coding_formats

<img width="983" alt="image" src="https://github.com/user-attachments/assets/1f970628-24e2-47e6-9955-351ce797bcd6">

**Comment by ldexterldesign** [comment]
Thanks for quick reply

Apologies for delay

Good thought about browser support! 🙄

```
ffprobe version 7.1 Copyright (c) 2007-2024 the FFmpeg developers
  built with Apple clang version 16.0.0 (clang-1600.0.26.4)
  configuration: --prefix=/opt/homebrew/Cellar/ffmpeg/7.1_3 --enable-shared --enable-pthreads --enable-version3 --cc=clang --host-cflags= --host-ldflags='-Wl,-ld_classic' --enable-ffplay --enable-gnutls --enable-gpl --enable-libaom --enable-libaribb24 --enable-libbluray --enable-libdav1d --enable-libharfbuzz --enable-libjxl --enable-libmp3lame --enable-libopus --enable-librav1e --enable-librist --enable-librubberband --enable-libsnappy --enable-libsrt --enable-libssh --enable-libsvtav1 --enable-libtesseract --enable-libtheora --enable-libvidstab --enable-libvmaf --enable-libvorbis --enable-libvpx --enable-libwebp --enable-libx264 --enable-libx265 --enable-libxml2 --enable-libxvid --enable-lzma --enable-libfontconfig --enable-libfreetype --enable-frei0r --enable-libass --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-libopenjpeg --enable-libspeex --enable-libsoxr --enable-libzmq --enable-libzimg --disable-libjack --disable-indev=jack --enable-videotoolbox --enable-audiotoolbox --enable-neon
  libavutil      59. 39.100 / 59. 39.100
  libavcodec     61. 19.100 / 61. 19.100
  libavformat    61.  7.100 / 61.  7.100
  libavdevice    61.  3.100 / 61.  3.100
  libavfilter    10.  4.100 / 10.  4.100
  libswscale      8.  3.100 /  8.  3.100
  libswresample   5.  3.100 /  5.  3.100
  libpostproc    58.  3.100 / 58.  3.100
Input #0, caf, from '/Users/ldexterldesign/Desktop/foo/compatible/attachments/103/3194.caf':
  Duration: 00:00:08.06, start: 0.000000, bitrate: 32 kb/s
  Stream #0:0: Audio: opus (opus / 0x7375706F), 48000 Hz, mono, fltp
```
... so files are .opus with .caf container

https://caniuse.com/opus - I tested in a bunch of browsers (inc. latest Safari, Chrome and Firefox) and no luck:

<img width="723" alt="Screenshot 2024-11-17 at 10 07 06" src="https://github.com/user-attachments/assets/14a43b68-cf91-4bf2-ac16-cb81d30cf262">

It seems the files will need to use .webm container

Perhaps converting .caf files to something compatible (i.e. .webm) could be part of `-o compatible` - if you agree then I can create a feature request as it would be a wonderfully useful addition and give life/longevity to any and all users who use voice messages (*a lot* I suspect; and increasing)

Yours hopefully

FMR `imessage-exporter -c compatible -f html -o .`

**Reply by ReagentX** [reply]
~~I will consider implementing that as part of #346.~~

**Reply by ReagentX** [reply, marked as answer]
Following up, the error you described above is due to disk permissions. If you do not copy the attachments, Safari attempts to read them from `~/Library/...` which fails with the following log:

<img width="285" alt="image" src="https://github.com/user-attachments/assets/1373d7b5-dd99-4cab-8fa6-09b0a040a2a3">

If you instead use `-c`, the file will preview correctly, assuming the browser supports the `x-caf` type:

<img width="513" alt="image" src="https://github.com/user-attachments/assets/7aa2b42f-65c7-44c2-a4f7-93391fe6e503">

<img width="715" alt="image" src="https://github.com/user-attachments/assets/d0348cfd-9024-4c0d-a566-c4b8a8c754e0">

**Comment by ldexterldesign** [comment]
Apologies for delay

> https://caniuse.com/opus - I tested in a bunch of browsers (inc. latest Safari, Chrome and Firefox) and no luck:

I obviously didn't test well enough because files do play in Safari(-only); true to your compatibility table screenshot 🙄

As you point out `-c compatible` or `-c efficient` must be used as `-c disabled` won't result in any files being copied to the file system so they can't/won't play - not a file permissions issue in this instance but thank you for thinking for me! 🤓

Hope this helps

Regards

PS great to see #346 progress!
